Add server side

brhane-bt opened the pull request:
Add id btnClose to the modal close button

In server folder

Add server side scripts connection.js for mongoose connection
Add task.js for model of the mytasks collection in todo document in
mongodb
Add server.js for http post/get/put and delete operations.

In js folder

Add http get operation to load tasks from database
Add onclick event listener function and http post to addMarker function
Add http delete and http get to load the latest to remove function
Add onclick event, new updated marker variable and http put to the Info
    function

brhane-bt committed:
n index.html

Add id btnClose to the modal close button

In server folder

Add server side scripts connection.js for mongoose connection
Add task.js for model of the mytasks collection in todo document in
mongodb
Add server.js for http post/get/put and delete operations.

In js folder

Add http get operation to load tasks from database
Add onclick event listener function and http post to addMarker function
Add http delete and http get to load the latest to remove function
Add onclick event, new updated marker variable and http put to the Info
    function

brhane-bt committed:
Add custom pop up information by tamene
Modify error taskPostSuccess to postSuccess

lbraun commented on js/index.js:
```diff
@@ -30,25 +33,72 @@ app.controller("IndexController", ["$scope", "$http", function($scope, $http) {
     $scope.currentMarker = marker;
     $("#modal").modal('show');
+    $("#modal").on('click','#btnClose',function(){
+      //console.log(marker[]);
```
Can this be removed?

brhane-bt replied:
I thought we agreed we needed the on click event listener. Didn't we?

lbraun replied:
I see, sorry for the confusion, just meant to remove line 54 (the commented out code). I like the event listener solution! 👍 💃 🎉

lbraun commented on js/index.js:
```diff
@@ -30,25 +33,72 @@ app.controller("IndexController", ["$scope", "$http", function($scope, $http) {
+    }
+  }
+  function loadTodos(response){
+    //console.log(response);
```
Can this be removed?

lbraun commented on js/index.js:
```diff
@@ -30,25 +33,72 @@ app.controller("IndexController", ["$scope", "$http", function($scope, $http) {
+    }
+  }
+  function postSuccess(response){
+    //console.log(response);
```
Can this be removed?

lbraun commented on js/index.js:
```diff
@@ -30,25 +33,72 @@ app.controller("IndexController", ["$scope", "$http", function($scope, $http) {
+  }
   $scope.info = function(index) {
-    $scope.currentMarker = $scope.markers[index];
+    var m = {
```
I've been taught that it is better practice to name variables as descriptively as possible, i.e. never use one letter variable names, because it makes your code more readable and maintainable. Could we just call this "marker" instead?

lbraun replied:
It's also a rule in the google style guide!
https://google.github.io/styleguide/jsguide.html#naming

brhane-bt replied:
Ops ... sorry. I know it is not recommended. I will do that.

lbraun commented on js/index.js:
```diff
@@ -30,25 +33,72 @@ app.controller("IndexController", ["$scope", "$http", function($scope, $http) {
+    //console.log(m);
+    $("#modal").on('click','#btnClose',function(){
+      //console.log($scope.markers[index]._id);
+      //console.log(marker[]);
```
Can the rest of these comments be removed?

lbraun commented on js/index.js:
```diff
@@ -30,25 +33,72 @@ app.controller("IndexController", ["$scope", "$http", function($scope, $http) {
 
+
+  $scope.remove = function(index) {
+    if (index != -1) {
```
👍 Does this fix the issue when deleting the last item?

brhane-bt replied:
I modified and solved it, but I guarantee it is not the best way. :-P

lbraun commented on js/index.js:
```diff
@@ -63,9 +113,5 @@ app.controller("IndexController", ["$scope", "$http", function($scope, $http) {
-      $scope.markers.splice(index, 1);
-    }
-  }
+
```
Why add this extra blank line?

lbraun commented on js/index.js:
```diff
@@ -30,25 +33,72 @@ app.controller("IndexController", ["$scope", "$http", function($scope, $http) {
+    $http.get(serverUrl).then(loadTodos,errorMessage);
+  }
+  function errorMessage(error){
+      console.log(error);
```
I think everywhere else we use 2 spaces for indentation. Should this be 2 as well?

lbraun commented on server/connection.js:
```diff
@@ -0,0 +1,10 @@
+var db = mongoose.connection;
+db.on("error", console.log.bind(console, "connection error."));
+db.on("open", function () {
+    console.log("Mongodb connected");
```
Use two spaces

lbraun commented on server/server.js:
```diff
@@ -0,0 +1,90 @@
+app.use(cors());
+// app.use(express.static(__dirname + '../html'));
+// app.use(express.static(__dirname + '../js'));
+// app.use(express.static(__dirname + '../node_modules'));
```
Remove this unused code

lbraun commented on server/server.js:
```diff
@@ -0,0 +1,90 @@
+app.use(bodyParser.json());
+
+app.post("/todo", function(req, res,next) {
+          task.create({"title": req.body.title,"dueDate": req.body.dueDate,"address": req.body.address,"lat": req.body.lat,"lng": req.body.lng}, function(err, task) {
```
Fix indentation

lbraun replied:
I don't think you need quotes around the keys, do you?

Maybe moving the hash to a different line would make this more readable too. What do you think?

lbraun commented on server/server.js:
```diff
@@ -0,0 +1,90 @@
+          });
+})
+
+app.delete("/todo/:id", function(req, res,next) {
```
Missing a space before `next`

lbraun commented on server/server.js:
```diff
@@ -0,0 +1,90 @@
+  }, function(err, result) {
+    if (err) {
+      res.status(500);
+
```
Remove this blank line for consistency?

lbraun commented on server/server.js:
```diff
@@ -0,0 +1,90 @@
+      res.status(500);
+
+      next("Internal server error.");
+    } else if (result.n == 0) {
```
Is result.n the number of results? That's good to know!

brhane-bt replied:
Yeah it is. There is also one unnecessary space isn't i :-P

lbraun commented on server/server.js:
```diff
@@ -0,0 +1,90 @@
+})
+app.put("/todo/:id", function(req, res,next) {
+  var id = '' + req.params.id;
+  console.log("here");
```
Remove this when done debugging.

lbraun commented on server/server.js:
```diff
@@ -0,0 +1,90 @@
+  if(stream){
+    var results = [];
+    stream.on('data', function(doc) {
+      //delete doc._id;
```
Delete this unused code

lbraun commented on server/server.js:
```diff
@@ -0,0 +1,90 @@
+    var results = [];
+    stream.on('data', function(doc) {
+      //delete doc._id;
+      delete doc.__v;
```
What is this `__v`? I've never heard of that? Maybe add a comment to help explain it?

brhane-bt replied:
As mentioned here, it is a property set on each document when first created by Mongoose.
http://mongoosejs.com/docs/guide.html#versionKey
That's why I am removing it. It is not necessary.

lbraun replied:
I see, interesting! Thanks for the link :)

lbraun commented on server/server.js:
```diff
@@ -0,0 +1,90 @@
+      res.send(results);
+    });
+  }
+  else{
```
Rewrite this like `} else {`

lbraun commented on server/task.js:
```diff
@@ -0,0 +1,11 @@
+    dueDate: {type: String, required: true},
+    address: {type: String, required: true},
+    lat: {type: Number, required:true },
+    lng: {type:Number, required:true}
```
Use two spaces for indentation

brhane-bt replied:
Really? I have always used tab for indentation. I will update that and pretty much everywhere too.

lbraun replied:
Tabs are fine too, but we should choose one way to do it and stick with it. Here you are using 4 spaces, not a tab, in any case. Maybe your editor is set to do 4 spaces by default?

lbraun commented on server/task.js:
```diff
@@ -0,0 +1,11 @@
+    title: {type: String, required: true},
+    dueDate: {type: String, required: true},
+    address: {type: String, required: true},
+    lat: {type: Number, required:true },
```
Add space before "true", remove space after

lbraun commented on server/task.js:
```diff
@@ -0,0 +1,11 @@
+var db = require("./connection");
+
+var task = db.model('mytasks', {
```
Maybe just call the model "tasks" instead of "mytasks"? 😄Who knows if they are really mine? 😉

lbraun reviewed (changes requested):
@brhanebt great work!!!!! All the logic looks good. I mostly added comments about formatting, etc. Let me know what you think, if you have any questions, or if you feel like my ideas are stupid 😆 I'd love to hear it.

brhane-bt committed:
Remove unncessary commented out statements
Modify remove last marker (Not efficient way)
Modify indentation
Rename mytasks to tasks
Modify popup message to include title and duedate only

lbraun commented on js/index.js:
```diff
@@ -28,44 +27,77 @@ app.controller("IndexController", ["$scope", "$http", function($scope, $http) {
+    var id = $scope.markers[index]._id;
+    $http.delete(serverUrl + id).then(function() {
+      if ($scope.markers.length == 1) {
+        window.location.reload();
```
We shouldn't have to do this to get the view to update, but it is a good fix for now. Maybe we can ask Oscar why the last element isn't disappearing.

lbraun commented on js/index.js:
```diff
@@ -28,44 +27,77 @@ app.controller("IndexController", ["$scope", "$http", function($scope, $http) {
+  function loadTodos(response) {
+    if (response.data.length > 0) {
+      $scope.markers = response.data;
+      for (var i = 0; i < $scope.markers.length; i++) {
```
Have you tried the jquery each() method? I find it nice for situations like these.

brhane-bt replied:
You are right, it looks nicer. I will change it. It's just that I read somewhere this is faster and I prefer simple for loop in cases.

lbraun commented on js/index.js:
```diff
@@ -28,44 +27,77 @@ app.controller("IndexController", ["$scope", "$http", function($scope, $http) {
+    if (response.data.length > 0) {
+      $scope.markers = response.data;
+      for (var i = 0; i < $scope.markers.length; i++) {
+        $scope.markers[i].message = "<b>Title: </b>" + $scope.markers[i].title + "<br/><b>Due Date: </b><b>" + $scope.markers[i].dueDate + "</b>";
```
This addresses @Tamenegelaye 's issue #3 doesn't it? Did you show him this?

lbraun commented on server/server.js:
```diff
@@ -0,0 +1,93 @@
+  });
+})
+app.delete("/todo/:id", function(req, res, next) {
+  var id = '' + req.params.id;
```
This doesn't come as a string already? That kind of surprises me.

brhane-bt replied:
You are right, that's a string already my bad :-)

lbraun reviewed (approved):
Looks GREAT Brhane. Let me know if you want to talk about how to fix the merge conflicts. What's left to do after this?!

lbraun commented on server/server.js:
```diff
@@ -0,0 +1,93 @@
+      results.push(doc);
+    }).on("error", function(err) {
+      res.status(500);
+      res.send(err);
```
Oh, should we use next() here instead?

brhanebt replied:
You think that's better? I just didn't understand what next is used for? Anyways, suggestion accepted.

brhane-bt replied:
You think that's better? I just didn't understand what next is used for? Anyways, suggestion accepted.

lbraun replied:
I don't totally understand it either, but I think it keeps the server from hanging on errored out requests by instructing it to go on to the next thing. [This page](https://expressjs.com/en/guide/error-handling.html) says it's important for garbage collection:
> Notice that when not calling “next” in an error-handling function, you are responsible for writing (and ending) the response. Otherwise those requests will “hang” and will not be eligible for garbage collection.

So maybe we should use it in addition to the res.send()? Or maybe it's not necessary since we are completing the request? I'm not sure 😕

brhane-bt committed:
Change For loop with For each
Add next when error

brhane-bt committed:
Modify the show function to add pop up after centering in index.js
Add Id for the leaflet element in index.html to identfy the map for
popup

brhane-bt committed:
Merge branch 'master' into Add-server-side
